### PR TITLE
Fin/function isomorphism

### PR DESCRIPTION
**Pull request by conal**
This PR adds an `Inverse`/`↔` for exponentials, augmenting `splitAt`/`join` (sums) and `remQuot`/`combine` (products). Sadly, this bundle needs extensionality for one direction.

Here are two alternative definitions for `fin→fun` that save a bit of repetition:

``` agda
fin→fun : ∀ {m n} → Fin (n ^ m) → (Fin m → Fin n)
fin→fun {suc m} {n} k i with remQuot {n} (n ^ m) k
fin→fun {suc m} _  zero   | j ,  _  = j
fin→fun {suc m} _ (suc i) | _ , n^m = fin→fun n^m i
```

``` agda
fin→fun : ∀ {m n} → Fin (n ^ m) → (Fin m → Fin n)
fin→fun {suc m} {n} k i = case i , remQuot {n} (n ^ m) k of λ
  { ( zero , j ,  _ )  → j
  ; (suc i , _ , n^m) → fin→fun n^m i }
```

Somehow I got extra commits (first through fifth) in the summary below. The actual changes below look correct.

### CODE REVIEW

**Comment by conal** [comment]
Any thoughts on this PR and particularly on functions vs vectors for the exponential isomorphism?

**Comment by MatthewDaggitt** [comment]
Hi @conal, apologies for letting this one slip off my immediate to-do list. I have to confess, given the two short-comings you've identified (the aesthetic loss and the dependency conundrum) I'm definitely on the side of keeping the functional version of the properties. 

If you wanted to you could then add your translation from vectors to functions in `Data.Vec.Base` in a future PR without any problems?

**Comment by conal** [comment]
Thanks for the reply, @MatthewDaggitt. Yes, I could define the isomorphism `Fin`/`Vec` isomorphism on top of the the `Fin`/`Fun` almost-isomorphism, thanks to `tabulate-cong` in `Data.Vec.Properties`, which converts extensional equality on functions to (regular) equality on vectors. I can't use `^↔→` itself for this purpose, since it depends on extensionality, but I can fairly conveniently use `fin→fun→fin`, `fun→fin→fun`, and `tabulate-cong`.

File: CHANGELOG.md [hunk 1]
```diff
@@ -489,11 +489,22 @@ Other minor changes
   ```
   and their corresponding algebraic substructures.
 
-* Added a new `Inverse` bundle in `Data.Fin.Properties`:
+* Added new functions in `Data.Fin.Base`:
+  ```
+  fin→fun : ∀ {m n} → Fin (n ^ m) → (Fin m → Fin n)
```
[the remark this answers is not in the document]
Comment (conal) [reply CHANGELOG.md]: Done. Thanks for the tip!

**Comment by conal** [comment]
Okay. I think I've addressed all suggestions (so far) about this PR. Assuming all checks pass, it's good to go as far as I'm concerned.

**Comment by MatthewDaggitt** [comment]
Hi @conal, apologies for the last minute bikeshedding, but I've just realised we're breaking naming conventions with `fin→fun→fin` etc., would you be opposed to me renaming `fin→fun→fin` to the standard `fin→fun-fun→fin`? Actually to be honest, having `→` in the name of a data-level function at all is a little bit non-standard. Would me changing `fin→fun` to `finToFun` break your heart?

**Comment by conal** [comment]
> Hi @conal, apologies for the last minute bikeshedding, but I've just realised we're breaking naming conventions with `fin→fun→fin` etc., would you be opposed to me renaming `fin→fun→fin` to the standard `fin→fun-fun→fin`? Actually to be honest, having `→` in the name of a data-level function at all is a little bit non-standard. Would me changing `fin→fun` to `finToFun` break your heart?

Hi @MatthewDaggitt. These changes are perfectly fine with me. Go for it.

**Comment by conal** [comment]
@MatthewDaggitt BTW, I'd be happy to make those changes myself if you'd like.

**Comment by MatthewDaggitt** [comment]
That's okay, I've already made most of them :+1:  Just need to push later on today.

**Comment by conal** [comment]
Cool. Thanks!

**Comment by MatthewDaggitt** [comment]
Test suite blocked on #1618

**Comment by MatthewDaggitt** [comment]
Tests are passing, so merging in. Thanks for your patience @conal !

**Pull request merged**